ext/drpttracer/drpttracer.c

```diff
@@ -0,0 +1,579 @@
+    uint64_t base_size = (uint64_t)((1UL << BASE_SIZE_SHIFT) + 1) * sysconf(_SC_PAGESIZE);
+    uint64_t base_mmap_size = base_size;
+    void *base =
+        dr_map_file(fd, &base_mmap_size, 0, NULL, DR_MEMPROT_READ | DR_MEMPROT_WRITE, 0);
+    if (base == NULL || base_mmap_size != base_size) {
+        ERRMSG("failed to mmap perf event for pt tracing\n");
+        dr_close_file(fd);
+        return NULL;
+    }
+
+    /* Initialize the header of the perf event file's mmap pages. */
+    struct perf_event_mmap_page *header = (struct perf_event_mmap_page *)base;
+    header->aux_offset = header->data_offset + header->data_size;
+    header->aux_size = (uint64_t)(1UL << AUX_SIZE_SHIFT) * sysconf(_SC_PAGESIZE);
+
+    /* Initialize the ring buffer of the perf event's auxiliary data. */
+    uint64_t aux_mmap_size = header->aux_size;
+    void *aux = dr_map_file(fd, &aux_mmap_size, header->aux_offset, NULL,
+                            DR_MEMPROT_READ | DR_MEMPROT_WRITE, 0);
+    if (aux == NULL || aux_mmap_size != header->aux_size) {
+        ERRMSG("failed to mmap aux for pt tracing\n");
+        dr_unmap_file(base, base_size);
+        dr_close_file(fd);
+        return NULL;
+    }
+
+    /* Alloc the handle and set it. */
+    pttracer_handle_t *handle =
+        (pttracer_handle_t *)dr_thread_alloc(drcontext, sizeof(pttracer_handle_t));
+    handle->fd = fd;
+    handle->base = base;
+    handle->base_size = base_size;
+    handle->header = header;
+    handle->aux = aux;
+    return handle;
+}
+
+/* This function is used to free a pttracer_handle_t.
+ */
+static void
+pttracer_handle_free(IN void *drcontext, pttracer_handle_t *handle)
+{
+    if (handle == NULL)
+        return;
+    if (handle->fd >= 0)
+        dr_close_file(handle->fd);
+    if (handle->aux != NULL)
+        dr_unmap_file(handle->aux, handle->header->aux_size);
+    if (handle->base != NULL)
+        dr_unmap_file(handle->base, handle->base_size);
+
+    dr_thread_free(drcontext, handle, sizeof(pttracer_handle_t));
+}
+
+/***************************************************************************
+ * INIT APIS
+ */
+
+static int drpttracer_init_count;
+
+DR_EXPORT
+bool
+drpttracer_init(void)
+{
+    /* handle multiple sets of init/exit calls */
+    int count = dr_atomic_add32_return_sum(&drpttracer_init_count, 1);
+    if (count > 1)
+        return true;
+    pt_shared_metadata_init(&pt_shared_metadata);
+    if (!(pt_perf_event_attr_init(true, false, &user_only_pe_attr) &&
+          pt_perf_event_attr_init(false, true, &kernel_only_pe_attr) &&
+          pt_perf_event_attr_init(true, true, &user_kernel_pe_attr))) {
+        ASSERT(false, "failed to init drpttracer's perf_event_attr");
+        drpttracer_exit();
+        return false;
+    }
+    return true;
+}
+
+DR_EXPORT
+void
+drpttracer_exit(void)
+{
+    /* handle multiple sets of init/exit calls */
+    int count = dr_atomic_add32_return_sum(&drpttracer_init_count, -1);
+    if (count != 0)
+        return;
```

Comment:
This looks like a no-op.

Maybe you wanted to do some cleanup after this?

Reply:
drpttracer doesn't have any data need to be cleanup when call its exit function. So it may looks like https://github.com/DynamoRIO/dynamorio/blob/15cde3aaa4e7ae0609813709d306ef36633384fb/ext/drutil/drutil.c#L145-L155
Do you agree I only add some similar comments.